### ISSUE TITLE
Old behavior of ctrl-c on 4.0b1 doesn't work with multi-line prompt

### ISSUE THREAD

**Issue by superobertking** [fish-shell/fish-shell#10935]
Version: fish, version 4.0b1-03a9f4
OS: Darwin 23.6.0 Darwin Kernel Version 23.6.0: Thu Sep 12 23:36:12 PDT 2024; root:xnu-10063.141.1.701.1~1/RELEASE_ARM64_T6020 arm64
$TERM: xterm-256color

Regarding this change of behavior mention in release note:

> ctrl-c during command input no longer prints ^C and a new prompt, but merely clears the command line. This restores the behavior from version 2.2. To revert to the old behavior, use bind ctrl-c __fish_cancel_commandline ([#10213](https://github.com/fish-shell/fish-shell/issues/10213)).

I was trying to revert back to the old behavior. However, setting `bind -M insert ctrl-c __fish_cancel_commandline` does not work well with multi-line prompt (I am using [starship](https://starship.rs)).

On old fish 3.7.1:
```
/tmp/testgit on ᚠ master
❯ asdfa^C
/tmp/testgit on ᚠ master
❯
```
On 4.0b1-03a9f4:
```
/tmp/testgit on ᚠ master
/tmp/testgit on ᚠ master
❯
```
The ^C was actually printed, but immediately got overwritten by the new repaint. It should repaint on a newline instead of resetting back to beginning of current line.

On 3.7.1, both the preset `cancel-commandline` and binding `__fish_cancel_commandline` work fine and they both keep the old command line.
```
❯ bind \cc
bind --preset -m insert \cc cancel-commandline repaint-mode   # either works fine
bind -m insert \cc __fish_cancel_commandline   # either works fine
```

BTW, one suggestion for the Release Note: for the bind commands, maybe also mention that vim binding users also need `-M insert` and `-M visual` for the binding to work.

**Comment by faho** [comment]
How about calling it "abandon-commandline" and just properly shipping it instead?

**Comment by krobelus** [comment]
`abandon-commandline` sounds good, and I yeah forgot docs (not sure what else?)

**Comment by krobelus** [comment]
We could also have left the behavior of `cancel-commandline` intact but my assumption is that it's a net positive.
I don't know if the "keep abandoned commands in scrollback" workflow is common

**Comment by Darkshadow2** [comment]
I use it all the time, to have a command that I'm referencing. Usually something long, and I'm just going to use a part of it and it'd be a pain to go through and delete everything else

**Comment by krobelus** [comment]
I see that makes sense. I would probably use the kill ring (`ctrl-k` etc.) for that, or a separate window with a text editor, since using the mouse to access scrollback feels too slow.

**Comment by Darkshadow2** [comment]
Well, I don't even scroll back in the window buffer, I usually just type part of the command then arrow up in the history until the command I want is shown, then hit ctrl-c and then reference it that way. So it's just right above what I'm currently typing.

Also, I'm not arguing against the change to have it just clear the line. It's just an old old habit, and I've been working on ways to do it differently.

**Comment by superobertking** [comment]
Thank you for the fix! That was really fast! The cancel behavior is now working flawlessly.

The only small episode I had was that I found when I put the bindings in config.fish, `bind --list-modes` only had "default". This is probably because I'm using `hybrid_bindings` ([mentioned in doc somewhere](https://github.com/fish-shell/fish-shell/issues/3556)) which doesn't get called until `fish_key_bindings` is called. Adding the `ctrl-c` bindings to `fish_user_key_bindings.fish` solves it for me.

Thank you all for the two year long efforts to transition to Rust!

**Comment by zanchey** [comment]
@krobelus is there more you want to do here for 4.0?

**Comment by krobelus** [comment]
better naming at least. Actually I think it would be preferrable to make the default binding something like `bind ctrl-c 'commandline ""'` (because that's the most obvious way). Today it's implemented as special readline command mostly due to historical reasons I think; and I don't expect anyone uses the `fish_cancel` event.
Then the `cancel-commandline` name would be free again for the old behavior but since it's been around for a while, we could also use a new name `abandon-commandline`, which is also a better description.

**Comment by faho** [comment]
So, what we could do here is either rename the things or back out #10213 entirely. I don't really mind either way, but we should do one.

My preferred naming would probably be "clear-commandline" for the new thing and "cancel-commandline" for the old way of leaving a `^C` marker, simply because that leaves cancel-commandline the same across versions.

I would also be fine with calling the new thing "cancel-commandline" and the old "abandon-commandline" - I really don't like the "-traditional" wording because it doesn't mean anything.

Depending on when we want to release 4.0 proper, I would just commit the "clear-commandline" version unless anyone comes up with anything better.

**Comment by krobelus** [comment]
On Sat, Jan 11, 2025 at 02:18:01AM -0800, Fabian Boehm wrote:
> So, what we could do here is either rename the things or back out #10213 entirely. I don't really mind either way, but we should do one.
> 
> My preferred naming would probably be "clear-commandline" for the new thing and "cancel-commandline" for the old way of leaving a `^C` marker, simply because that leaves cancel-commandline the same across versions.
> 
> I would also be fine with calling the new thing "cancel-commandline" and the old "abandon-commandline" - I really don't like the "-traditional" wording because it doesn't mean anything.

dubbing the current behavior `clear-commandline` and naming the old behavior `cancel-commandline` again sounds good to me (probably not too many people have copied it to  their config but stability is still good).

> Depending on when we want to release 4.0 proper, I would just commit the "clear-commandline" version unless anyone comes up with anything better.

Not sure what you mean but I would not change the default back. It took quite a lot of effort, head-scratching and months of hesitation; I'd rather not have that destroyed by a [low-effort change again](https://github.com/fish-shell/fish-shell/commit/5f849d02646b60943557b00b51a0e592c2387959).

> Well, I don't even scroll back in the window buffer, I usually
> just type part of the command then arrow up in the history until
> the command I want is shown, then hit ctrl-c and then reference it
> that way. So it's just right above what I'm currently typing.
> 
> Also, I'm not arguing against the change to have it just clear the
> line. It's just an old old habit, and I've been working on ways to
> do it differently.

Ok. FWIW we also have `alt-#` for a similar purpose (though that one could use some love),
`ctrl-r` operating on the current line (getting it exactly when you need it, no need to prepare it ahead of time),
and  in latest master, multi-line autosuggestions.

> I use it all the time, to have a command that I'm referencing. Usually
> something long, and I'm just going to use a part of it and it'd be
> a pain to go through and delete everything else

Would be interesting to see an example. If the desired part is a single token, then `alt-.` usually works well.

One idea (advanced workflow, for huge commands) is `alt-enter`, `ctlr-r`, recall the desired  command, `alt-d` to kill what you want, `ctrl-z` remove the command again, `ctrl-y`

**Comment by faho** [comment]
> Not sure what you mean but I would not change the default back

What I mean is I have a slight preference for this plan:

Rename the new behavior to "clear-commandline" and change the bindings to match. Anyone who uses "cancel-commandline" gets the old behavior.

Basically instead of changing what "cancel-commandline" means, achieve the same outcome as we have now (ctrl-c clears the commandline) by introducing a new bind function and using it. But leave the old bind function intact so people who prefer that behavior can use it, and can use it across fish versions.

**Comment by krobelus** [comment]
Yeah I agree. Changing `cancel-commandline` was quite reckless,
in that change I mainly focused on matching pre-existing behavior;
(though again the `fish_cancel` event should probably be phased out)

**Comment by superobertking** [comment]
Just updated my fish to HEAD. I guess the [changelog doc](https://github.com/fish-shell/fish-shell/blob/945a53557038062eeb0cae551f8014cef5f3310f/CHANGELOG.rst?plain=1#L114) and [this function](https://github.com/fish-shell/fish-shell/blob/945a53557038062eeb0cae551f8014cef5f3310f/share/functions/__fish_cancel_commandline.fish#L3) also needs to remove the -traditional suffix?

**Comment by faho** [comment]
(fixed the function in 4.0 and master, we're probably gonna merge the CHANGELOG back once that's out, so I haven't changed that in master)

**Comment by rickalex21** [comment]
> Well, I don't even scroll back in the window buffer, I usually just type part of the command then arrow up in the history until the command I want is shown, then hit ctrl-c and then reference it that way. So it's just right above what I'm currently typing.
> 
> Also, I'm not arguing against the change to have it just clear the line. It's just an old old habit, and I've been working on ways to do it differently.

I agree with this, I do the same thing. How can I get the old behavior of ctrl-c? IMO Ctrl-c should have been left alone since this is the way most programs work and an option for a user to override. It should not be the default.

**Comment by faho** [comment]
@rickalex21 As the [release notes say](https://fishshell.com/docs/current/relnotes.html#notable-backwards-incompatible-changes):

> ctrl-c now calls a new bind function called clear-commandline. The old behavior, which leaves a “^C” marker, is available as cancel-commandline

So: `bind ctrl-c cancel-commandline`.

**Comment by rickalex21** [comment]
@superobertking Thank You for your help that works.
